### ISSUE TITLE
mz_adapter::util: halting process: updating cluster status cannot fail: Catalog(Error { kind: Stash(StashError { inner: Transaction(Commit(StashError { inner: Postgres(Error { kind: Closed, cause: None }) })) }) })

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#17862]
### What version of Materialize are you using?

v0.45.0-dev (1b5b7fe6c)

### How did you install Materialize?

Docker image

### What is the issue?

This just failed in Nightly and I do not think it has ever been seen before. Could it be related to the recent refactor of CRDB error handling? @mjibson 

https://buildkite.com/materialize/nightlies/builds/1868#01869571-bbf3-4b88-9f37-b11f93bc6f5b

ci-regexp: Transaction\(Commit\(StashError

### Relevant log output

```source
zippy-materialized-1     | environmentd: 2023-02-28T00:45:06.786550Z  WARN deadpool.postgres: Connection error: connection closed    
zippy-materialized-1     | environmentd: 2023-02-28T00:45:06.786554Z ERROR mz_stash::postgres: postgres stash connection error: connection closed
zippy-materialized-1     | environmentd: 2023-02-28T00:45:06.786599Z ERROR mz_stash::postgres: postgres stash connection error: connection closed
zippy-materialized-1     | environmentd: 2023-02-28T00:45:06.786642Z ERROR mz_stash::postgres: postgres stash connection error: connection closed
zippy-materialized-1     | environmentd: 2023-02-28T00:45:06.786657Z  WARN deadpool.postgres: Connection error: connection closed    
zippy-materialized-1     | environmentd: 2023-02-28T00:45:06.786593Z ERROR mz_stash::postgres: postgres stash connection error: error communicating with the server: Connection reset by peer (os error 104)
zippy-materialized-1     | environmentd: 2023-02-28T00:45:06.786732Z  WARN mz_adapter::util: halting process: updating cluster status cannot fail: Catalog(Error { kind: Stash(StashError { inner: Transaction(Commit(StashError { inner: Postgres(Error { kind: Closed, cause: None }) })) }) })
```

**Comment by mjibson** [comment]
Yup, this is exactly that. I am already working on a PR to make this auto retry so this doesn't happen.

**Comment by mjibson** [comment]
Closed by #17920
